### PR TITLE
[reflection] Throw TLE for Type.GetType("", ...) (Fixes #59664)

### PR DESCRIPTION
**Pull request by lambdageek**
Cherrypick #5751 to `2017-08`

Fixes [#59664](https://bugzilla.xamarin.com/show_bug.cgi?id=59664)

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -4446,9 +4442,11 @@ ves_icall_System_Reflection_Assembly_InternalGetType (MonoReflectionAssemblyHand
 		goto fail;
 
 	/*g_print ("requested type %s in %s\n", str, assembly->assembly->aname.name);*/
-	if (!mono_reflection_parse_type (str, &info)) {
+	MonoError parse_error;
+	if (!mono_reflection_parse_type_checked (str, &info, &parse_error)) {
 		g_free (str);
 		mono_reflection_free_type_info (&info);
+		mono_error_cleanup (&parse_error);
```
Comment (kumpera) [inline mono/metadata/icall.c]: We should concat the parse error to the mono_error_set_argument below, shouldn't we?

File: mono/metadata/icall.c [hunk 2]
```diff
@@ -1443,30 +1443,24 @@ ves_icall_System_Type_internal_from_name (MonoStringHandle name,
 {
 	error_init (error);
 	MonoTypeNameParse info;
-	gboolean parsedOk;
+	gboolean free_info = FALSE;
```
Comment (kumpera) [inline mono/metadata/icall.c]: I'm wondering if we should adopt the same idiom of MonoError with MonoTypeNameParse.
IE, we always init it to a state that signals no free'ing is needed.

That would simplify a lot of the usage within the runtime.

Comment (lambdageek) [reply mono/metadata/icall.c]: You know what's funny? we have `MONO_API` functions that consume a `MonoTypeNameParse*` but none that create one, and the underlying struct isn't exposed. So there's no way to actually use, for example, `mono_reflection_parse_type` without access to `metadata-internals.h`.

So yea, I can fixup all the callers on `master` (This PR is for `2017-08`).

File: mono/metadata/icall.c (hunk 1 above)
Comment (lambdageek) [reply mono/metadata/icall.c]: There's no useful message in there yet.

**Comment by marek-safar** [comment]
I think this is below the bar change, closing for this branch

**Pull request closed without merging**